Maestro API with support for release pipelines

JohnTortugo opened the pull request:
Closes: https://github.com/dotnet/arcade/issues/1769

Created a new version of the API to add support for management of release pipelines and also associating release pipelines with channels.

ChadNedzlek commented on src/Maestro/Maestro.Data/Migrations/BuildAssetRegistryContextModelSnapshot.cs:
```diff
@@ -607,12 +607,12 @@ protected override void BuildModel(ModelBuilder modelBuilder)
                         .WithMany("ChannelReleasePipelines")
                         .HasForeignKey("ReleasePipelineId")
-                        .OnDelete(DeleteBehavior.Restrict);
+                        .OnDelete(DeleteBehavior.Cascade);
```
It looks like Cascade again... talk with @alexperovich

JohnTortugo replied:
I think I got it right this time :-)

ChadNedzlek commented on src/Maestro/Maestro.Web/Api/v2019_01_10/Controllers/BuildsController.cs:
```diff
@@ -0,0 +1,177 @@
+{
+    [Route("builds")]
+    [ApiVersion("2019-01-10")]
+    public class BuildsController : Controller
```
These controllers seem to be a duplication of the entire class from the previous version. Why redefine everything?  There's a lot of duplicates code/logic here that we shouldn't be duplicating

ChadNedzlek commented:
Most of this change seem to be duplicating stuff we shouldn't have to duplicate.  The entire web layer is copy pasted...

mmitche commented:
I don't think we need a new version of the API unless we make a breaking change to the existing API.  Shouldn't the release pipelines only be an add-on to the existing API?

alexperovich commented:
Yea, we shouldn't need a new API version unless there are breaking changes, and if we did need a new version all of the controllers should just inherit from the previous rather than duplicating the code.

JohnTortugo commented:
There is no breaking change, but there is a change to the format of the output that's way I thought creating a new version would be better. I duplicated the files because I thought I had seen it done that way somewhere.. maybe I'm just crazy haha :-/

Patching the current version is good enough for me! Thanks!

alexperovich commented:
Changing the output format in a way that will break callers expecting the old format is also a breaking change.



Sent from my Verizon, Samsung Galaxy smartphone


-------- Original message --------
From: Divino César <notifications@github.com>
Date: 1/11/19 12:05 PM (GMT-08:00)
To: dotnet/arcade-services <arcade-services@noreply.github.com>
Cc: Alex Perovich <Alex.Perovich@microsoft.com>, Mention <mention@noreply.github.com>
Subject: Re: [dotnet/arcade-services] New version of Maestro API with support for release pipelines (#72)


There is no breaking change, but there is a change to the format of the output that's way I thought creating a new version would be better. I duplicated the files because I thought I had seen it done that way somewhere.. maybe I'm just crazy haha :-/

Patching the current version is good enough for me! Thanks!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Farcade-services%2Fpull%2F72%23issuecomment-453640701&data=02%7C01%7CAlex.Perovich%40microsoft.com%7C3e75d28e62ce4b7e255708d678001601%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636828339099524706&sdata=7JEPOCc5cFBjptUH7pl1fnGuJzdV8C1ToP36GllRhI4%3D&reserved=0>, or mute the thread<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKE3IuAu7kVJD24iYGiDUtFEQDHP3DKQks5vCO5ygaJpZM4Z6wPc&data=02%7C01%7CAlex.Perovich%40microsoft.com%7C3e75d28e62ce4b7e255708d678001601%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636828339099524706&sdata=Trffg%2FJweXEz6gGy1%2FQFLf9Ln2o8up1LiWwivGY0dz4%3D&reserved=0>.

mmitche commented:
What parts of the output are changing?

ChadNedzlek commented:
If it's adding extra outputs (or even optional inputs), it's fine to not update, since a caller expecting only the old version would get the behavior they expected.  It's when you remove something or add a new required input, that you need a new version of everything, since that means old callers would suddenly get incorrect behavior. :-)

JohnTortugo commented:
> What parts of the output are changing?

when returning a channel I also return the list of pipelines associated with a channel.

mmitche commented:
That's fine. Data addition is okay as a json deserializer will generally ignore fields it doesn't know about.

JohnTortugo commented:
I just pushed new changes:

- Changed the approach to just patch the current API to add support for the management of pipelines and link the pipelines to channels.
- Adjusted the Cascade -> Restrict thing
- Renamed the ReleasePipeline*s* table

PTAL

alexperovich commented on src/Maestro/Maestro.Web/Api/v2018_07_16/Controllers/ChannelsController.cs:
```diff
@@ -40,7 +40,11 @@ public IActionResult Get(string classification = null)
             }
 
-            List<Channel> results = query.AsEnumerable().Select(c => new Channel(c)).ToList();
+            List<Channel> results = query.Include(ch => ch.ChannelReleasePipelines)
```
Please don't include all this information in this API. This is a "list" api, it shouldn't grab all the data for each item.

JohnTortugo replied:
You mean only remove it on this specific method, right?

alexperovich replied:
yea, the apis that return a single channel can include it. This one should not.

alexperovich commented on src/Maestro/Maestro.Web/Api/v2018_07_16/Controllers/ChannelsController.cs:
```diff
@@ -61,16 +68,27 @@ public async Task<IActionResult> GetChannel(int id)
 
         [HttpDelete("{id}")]
         [SwaggerResponse((int)HttpStatusCode.OK, Type = typeof(Channel))]
+        [SwaggerResponse((int)HttpStatusCode.BadRequest)]
```
These `SwaggerResponse` attributes should not be necessary. Why are you adding them?

JohnTortugo replied:
I'll remove them.

alexperovich commented on src/Maestro/Maestro.Web/Api/v2018_07_16/Controllers/ChannelsController.cs:
```diff
@@ -61,16 +68,27 @@ public async Task<IActionResult> GetChannel(int id)
+            {
+                return BadRequest(
+                    new ApiError($"The channel with id '{id}' has '{channel.ChannelReleasePipelines.Count()}' " +
+                    $"release pipeline(s) attached to it. Remove that relation first."));
```
>Remove that relation first. [](start = 58, length = 27)

This should probably say: `Detach those release pipelines(s) first.`

alexperovich commented on src/Maestro/Maestro.Web/Api/v2018_07_16/Controllers/ChannelsController.cs:
```diff
@@ -130,5 +148,71 @@ public async Task<IActionResult> AddBuildToChannel(int channelId, int buildId)
+            if (pipeline.ChannelReleasePipelines != null &&
+                pipeline.ChannelReleasePipelines.Any(existingPipelineChannel => existingPipelineChannel.ChannelId == channelId))
+            {
+                return StatusCode((int)HttpStatusCode.Created);
```
This should probably return statuse code NotModified in this case.

alexperovich commented on src/Maestro/Maestro.Web/Api/v2018_07_16/Controllers/ChannelsController.cs:
```diff
@@ -130,5 +148,71 @@ public async Task<IActionResult> AddBuildToChannel(int channelId, int buildId)
+            // If pipeline is not in the channel, nothing to do
+            if (pipeline == null)
+            {
+                return StatusCode((int)HttpStatusCode.OK);
```
This should be NotModified because it didn't change anything.

JohnTortugo commented:
Anything else I should address?
